vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/algebricksalternatives/MoveFreeVariableOperatorOutOfSubplanRule.java

```diff
@@ -0,0 +1,178 @@
+
+            //Make sure we are looking at subplan with a scan/join
+            if (op2.getInputs().size() != 1 || !descOrSelfIsScanOrJoin(op2)) {
+                return false;
+            }
+            boolean notApplicable = false;
+
+            //Get its used variables
+            Set<LogicalVariable> used = new HashSet<LogicalVariable>();
+
+            //not movable if the operator is not an assign or subplan
+            //Might be helpful in the future for other operations in the future
+            if (op2.getOperatorTag() == LogicalOperatorTag.ASSIGN) {
+                VariableUtilities.getUsedVariables(op2, used);
+            } else if (op2.getOperatorTag() == LogicalOperatorTag.SUBPLAN) {
+                // Nested plan must have an aggregate root.
+                ListIterator<ILogicalPlan> subplansIter = ((SubplanOperator) op2).getNestedPlans().listIterator();
+                ILogicalPlan plan = null;
+                while (subplansIter.hasNext()) {
+                    plan = subplansIter.next();
+                }
+                if (plan == null) {
+                    return false;
+                }
+                if (plan.getRoots().size() != 1) {
+                    return false;
+                }
+                ILogicalOperator op3 = plan.getRoots().get(0).getValue();
+                if (op3.getOperatorTag() != LogicalOperatorTag.AGGREGATE) {
+                    return false;
+                }
+                // Used variables do not include ones created in the subplan.
+                VariableUtilities.getUsedVariables(op2, used);
+                Set<LogicalVariable> subplanProducedAndDown = new HashSet<LogicalVariable>();
+                VariableUtilities.getProducedVariablesInDescendantsAndSelf(op3, subplanProducedAndDown);
+                used.removeAll(subplanProducedAndDown);
+            } else {
+                notApplicable = true;
+            }
+
+            //Make sure that all of its used variables come from outside
+            for (LogicalVariable var : used) {
+                if (!free.contains(var)) {
+                    notApplicable = true;
+                }
+            }
+
+            if (notApplicable) {
+                op2Ref = op2.getInputs().get(0);
+            } else {
+                //Make the input of op2 be the input of op1
+                op2Ref.setValue(op2.getInputs().get(0).getValue());
+
+                //Make the outside of the subplan the input of op2
+                Mutable<ILogicalOperator> outsideRef = op2.getInputs().get(0);
+                outsideRef.setValue(op0.getInputs().get(0).getValue());
+
+                //Make op2 the input of the subplan
+                Mutable<ILogicalOperator> op2OutsideRef = op0.getInputs().get(0);
+                op2OutsideRef.setValue(op2);
+
+                return true;
+            }
+
+        }
+        return false;
+    }
```

Comment:
We have a very similar rule to this in Hyracks (Movefreevariableoperatoroutofsubplanrule). Is there a reason to have these rules separated?